### PR TITLE
feat(index): try being resilient on "changes"

### PR DESCRIPTION
**Pull request by Haroenv**
There doesn't seem to be a builtin way of retrying on .changes pouchdb (https://github.com/pouchdb/pouchdb/issues/7209)

I tried this out, but I have not been able to make this fail on my local machine, because I don't manage to set up a MITM where i can fake a wrong response.

Only  approve this if you have tried it out and see no issues

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -313,20 +333,27 @@ async function watch(stateManager, mainIndex) {
...
       });
     });
-    listener.on('error', reject);
+
+    listener.on('error', err => {
+      sentry.report(err, {
+        retry: true,
+      });
+      log.info('🛰  Retrying watch', retry + 1);
+      watch(stateManager, mainIndex, retry + 1, err);
```
Comment (bodinsamuel) [inline src/index.js]: need async/await probably

File: src/index.js [hunk 2]
```diff
@@ -313,20 +333,27 @@ async function watch(stateManager, mainIndex) {
...
+    });
   });
 }
 
-async function error(err) {
-  sentry.report(err);
+async function error(err, extra) {
+  sentry.report(err, extra);
```
Comment (bodinsamuel) [inline src/index.js]: This function is never called directly, it can not have extra.

File: src/index.js [hunk 3]
```diff
@@ -250,9 +257,18 @@ async function replicate(stateManager, mainIndex) {
...
+    let lastError = undefined;
+    listener.on('error', err => {
+      lastError = err;
+      listener.cancel();
+    });
 
     changesConsumer.drain(() => {
+      if (lastError) {
+        log.info('🐌  Retrying replicate', retry + 1);
+        replicate(stateManager, mainIndex, retry + 1, lastError);
```
Comment (bodinsamuel) [inline src/index.js]: need async/await probably

File: src/index.js [hunk 4]
```diff
@@ -250,9 +257,18 @@ async function replicate(stateManager, mainIndex) {
         listener.cancel();
       }
     });
-    listener.on('error', reject);
+    let lastError = undefined;
+    listener.on('error', err => {
+      lastError = err;
+      listener.cancel();
```
Comment (bodinsamuel) [inline src/index.js]: On cancel does the `.drain()` is called ?

File: src/index.js (hunk 1 above)
Comment (Haroenv) [reply src/index.js]: Good call, needs a return

File: src/index.js (hunk 4 above)
Comment (Haroenv) [reply src/index.js]: We want to early cancel if eg 2/10 fails, we don’t want to continue with the other ones, or else it’ll skip one & be complicated

File: src/index.js (hunk 2 above)
Comment (Haroenv) [reply src/index.js]: The reject should bubble all the way up to main, no?

Comment (bodinsamuel) [reply src/index.js]: true, I didn't saw your second params on the reject. Not sure it work though `reject` only accept one param

**Pull request closed without merging**